lib/Sema/TypeCheckDistributed.cpp

```diff
@@ -127,44 +149,95 @@ bool swift::checkDistributedFunction(FuncDecl *func, bool diagnose) {
-    // TODO: we want to lift this restriction but it is tricky
-    ctor->diagnose(diag::distributed_actor_local_init_explicitly_defined)
-        .fixItRemove(SourceRange(ctor->getStartLoc(), decl->getStartLoc()));
-    // TODO: we should be able to allow this, but then we need to inject
-    //       code or force users to "do the right thing"
+  if (!swift::ensureDistributedModuleLoaded(ctor))
     return;
+
+  // === Designated initializers must accept exactly one ActorTransport
+  auto &C = ctor->getASTContext();
+  auto module = ctor->getParentModule();
+
+  SmallVector<ParamDecl*, 2> transportParams;
+  int transportParamsCount = 0;
+  auto protocolDecl = C.getProtocol(KnownProtocolKind::ActorTransport);
+  auto protocolTy = protocolDecl->getDeclaredInterfaceType();
+
+  for (auto param : *ctor->getParameters()) {
+    auto paramTy = ctor->mapTypeIntoContext(param->getInterfaceType());
+    auto conformance = TypeChecker::conformsToProtocol(paramTy, protocolDecl, module);
+
+    if (paramTy->isEqual(protocolTy) || !conformance.isInvalid()) {
+      transportParamsCount += 1;
+      transportParams.push_back(param);
+    }
   }
 
-  if (ctor->isDistributedActorResolveInit()) {
-    // It is illegal for users to attempt defining a resolve initializer;
-    // Suggest removing it entirely, there is no way users can implement this init.
-    ctor->diagnose(diag::distributed_actor_init_resolve_must_not_be_user_defined)
-        .fixItRemove(SourceRange(ctor->getStartLoc(), decl->getStartLoc()));
+  // missing transport parameter
+  if (transportParamsCount == 0) {
+    ctor->diagnose(diag::distributed_actor_designated_ctor_missing_transport_param,
+                   ctor->getName());
+    // TODO(distributed): offer fixit to insert 'transport: ActorTransport'
     return;
   }
 
-  // All user defined initializers on distributed actors must be 'convenience'.
-  //
-  // The only initializer that is allowed to be designated is init(transport:)
-  // which we synthesize on behalf of a distributed actor.
-  //
-  // When checking ctor bodies we'll check
-  if (!ctor->isConvenienceInit()) {
-    ctor->diagnose(diag::distributed_actor_init_user_defined_must_be_convenience,
-                   ctor->getName())
-        .fixItInsert(ctor->getConstructorLoc(), "convenience ");
+  // ok! We found exactly one transport parameter
+  if (transportParamsCount == 1)
     return;
-  }
+
+  // TODO(distributed): could list exact parameters
+  ctor->diagnose(diag::distributed_actor_designated_ctor_must_have_one_transport_param,
+                 ctor->getName(), transportParamsCount);
+}
+
+// ==== ------------------------------------------------------------------------
+
+void TypeChecker::checkDistributedActor(ClassDecl *decl) {
+  // ==== Ensure the _Distributed module is available,
+  // without it there's no reason to check the decl in more detail anyway.
+  if (!swift::ensureDistributedModuleLoaded(decl))
+    return;
```

Comment:
If this is _definitely_ called for every distributed-actor decl, can we go through and remove calls to this in later stages of the compiler? To me, it seems like we're excessively checking for it (even though it's a cached answer), because my understanding is that we just need to do it once for every distributed actor decl.

Reply:
You're likely right -- I'd hope to check only once.

I'm not sure about extensions...? I'll have to double check with an extension with a dist func inside

Reply:
Ok, removed a bunch of those!